synapse/federation/transport/server/_base.py

```diff
@@ -82,10 +82,15 @@ async def authenticate_request(self, request, content):
 
         for auth in auth_headers:
             if auth.startswith(b"X-Matrix"):
-                (origin, key, sig) = _parse_auth_header(auth)
+                (origin, key, sig, destination) = _parse_auth_header(auth)
                 json_request["origin"] = origin
                 json_request["signatures"].setdefault(origin, {})[key] = sig
 
+                # if the origin_server sent a destination along it needs to match our own server_name
+                if destination is not None and destination != self.server_name:
+                    raise AuthenticationError(
+                        400, "Destination mismatch in auth header", Codes.UNAUTHORIZED
```

Comment:
Could we use `HTTPStatus.BAD_REQUEST` here instead of 400? Or should this be 401, `HTTPStatus.UNAUTHORIZED`?

(Is unauthorized the right outcome here? As I understand it, we could confirm that this is from the originating server---it's just that this request isn't meant for us?)

Reply:
Yep, we received a request that isn't meant for us. This means that the authorization was completed correctly, so 400 feels better than 401.

I'll use `HTTPStatus.BAD_REQUEST` though ^^